### PR TITLE
Clean up left over uber principal resources for AWS

### LINKED ISSUE #2360: [BUG]: `create-uber-principal` command have to cleanup/resumability
### Current Behavior

`create-uber-principal` commands creates residue roles/policies/etc.

### Expected Behavior

When the commands fails, we have to cleanup any residue created by those commands. 

### Cloud

AWS, Azure

### PR DESCRIPTION
**Pull request by FastLee**
Partially resolves #2360 (AWS)

### CODE REVIEW

**Review by nfx — APPROVED** [review]
lgtm

File: src/databricks/labs/ucx/aws/access.py [hunk 1]
```diff
@@ -296,18 +296,24 @@ def get_instance_profile(self, instance_profile_name: str) -> AWSInstanceProfile
...
         return AWSInstanceProfile(instance_profile_arn)
 
-    def _create_uber_instance_profile(self, iam_role_name):
-        self._aws_resources.create_migration_role(iam_role_name)
-        self._aws_resources.create_instance_profile(iam_role_name)
+    def _create_uber_instance_profile(self, iam_role_name: str, iam_policy_name: str, s3_paths: set[str]):
         # Add role to instance profile - they have the same name
-        self._aws_resources.add_role_to_instance_profile(iam_role_name, iam_role_name)
+        if (
+            not (self._aws_resources.create_migration_role(iam_role_name))
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/aws/access.py:302]: Will these methods throw an exception? If so, they are not caught now

File: src/databricks/labs/ucx/aws/access.py [hunk 2]
```diff
@@ -296,18 +296,24 @@ def get_instance_profile(self, instance_profile_name: str) -> AWSInstanceProfile
...
+            or not (self._aws_resources.create_instance_profile(iam_role_name))
+            or not (self._aws_resources.add_role_to_instance_profile(iam_role_name, iam_role_name))
+            or not (
+                self._aws_resources.put_role_policy(
+                    iam_role_name, iam_policy_name, s3_paths, self._aws_account_id, self._kms_key
+                )
+            )
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/aws/access.py:309]: Remove redundant parenthesis:

```suggestion
            not self._aws_resources.create_migration_role(iam_role_name)
            or not self._aws_resources.create_instance_profile(iam_role_name)
            or not self._aws_resources.add_role_to_instance_profile(iam_role_name, iam_role_name)
            or not self._aws_resources.put_role_policy(
                    iam_role_name, iam_policy_name, s3_paths, self._aws_account_id, self._kms_key
                )
```

File: src/databricks/labs/ucx/aws/access.py [hunk 3]
```diff
@@ -359,8 +361,12 @@ def create_uber_principal(self, prompts: Prompts):
             self._update_sql_dac_with_instance_profile(iam_instance_profile, prompts)
             logger.info(f"Cluster policy \"{cluster_policy.name}\" updated successfully")
         except PermissionError:
+            logger.error(f"Failed to assign instance profile to cluster policy {iam_role_name}")
             self._aws_resources.delete_instance_profile(iam_role_name, iam_role_name)
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/aws/access.py:365]: Does this also delete the created roles?

File: src/databricks/labs/ucx/aws/access.py [hunk 4]
```diff
@@ -359,8 +361,12 @@ def create_uber_principal(self, prompts: Prompts):
...
 
+    def _delete_iam_role(self, role_name):
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/aws/access.py:367]: Where is this used?

**Review by JCZuurmond — COMMENTED** [review]
See comments

File: src/databricks/labs/ucx/aws/access.py (hunk 1 above)
Comment (FastLee) [reply src/databricks/labs/ucx/aws/access.py:302]: They are returning a boolean success/failure

File: src/databricks/labs/ucx/aws/access.py (hunk 2 above)
Comment (FastLee) [reply src/databricks/labs/ucx/aws/access.py:309]: OK

File: src/databricks/labs/ucx/aws/access.py (hunk 3 above)
Comment (FastLee) [reply src/databricks/labs/ucx/aws/access.py:365]: It does

File: src/databricks/labs/ucx/aws/access.py (hunk 4 above)
Comment (FastLee) [reply src/databricks/labs/ucx/aws/access.py:367]: Removed

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**